src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerStatsService.java

```diff
@@ -26,22 +49,81 @@ public KeyworkerStatsDto getStatsForStaff(Long staffId, String prisonId, LocalDa
         Validate.notNull(fromDate, "fromDate");
         Validate.notNull(toDate, "toDate");
 
-        List<CaseNoteUsageDto> usageCounts =
-                nomisService.getCaseNoteUsage(Arrays.asList(staffId), "KA", null, fromDate, toDate);
+        final LocalDateTime nextEndDate = toDate.atStartOfDay().plusDays(1);
+
+        List<OffenderKeyworker> applicableAssignments = repository.findByStaffIdAndPrisonId(staffId, prisonId).stream()
+                .filter(kw ->
+                        kw.getAssignedDateTime().compareTo(nextEndDate) < 0 &&
+                                (kw.getExpiryDateTime() == null || kw.getExpiryDateTime().compareTo(fromDate.atStartOfDay()) >= 0))
+                        .collect(Collectors.toList());
 
-        Map<String, Integer> usageGroupedBySubType =  usageCounts.stream()
-                .collect(Collectors.groupingBy(CaseNoteUsageDto::getCaseNoteSubType,
-                        Collectors.summingInt(CaseNoteUsageDto::getNumCaseNotes)));
+        List<String> prisonerNosList = applicableAssignments.stream().map(OffenderKeyworker::getOffenderNo).distinct().collect(Collectors.toList());
 
-        Integer entryCount = usageGroupedBySubType.get("KA");
-        Integer sessionCount = usageGroupedBySubType.get("KS");
+        if (!prisonerNosList.isEmpty()) {
+            List<CaseNoteUsagePrisonersDto> usageCounts =
+                    nomisService.getCaseNoteUsageForPrisoners(prisonerNosList, KEYWORKER_CASENOTE_TYPE, null, fromDate, toDate);
 
+            Map<String, Integer> usageGroupedBySubType = usageCounts.stream()
+                    .collect(Collectors.groupingBy(CaseNoteUsagePrisonersDto::getCaseNoteSubType,
+                            Collectors.summingInt(CaseNoteUsagePrisonersDto::getNumCaseNotes)));
+
+            Integer sessionCount = usageGroupedBySubType.get(KEYWORKER_SESSION_SUB_TYPE);
+            int sessionsDone = sessionCount != null ? sessionCount : 0;
+
+            int projectedKeyworkerSessions = getProjectedKeyworkerSessions(applicableAssignments, staffId, prisonId, fromDate, nextEndDate);
+            final BigDecimal complianceRate = getComplianceRate(sessionsDone, projectedKeyworkerSessions);
+
+            Integer entryCount = usageGroupedBySubType.get(KEYWORKER_ENTRY_SUB_TYPE);
+
+            return KeyworkerStatsDto.builder()
+                    .projectedKeyworkerSessions(projectedKeyworkerSessions)
+                    .complianceRate(complianceRate)
+                    .caseNoteEntryCount(entryCount != null ? entryCount : 0)
```

Comment:
This is done differently to sessionCount (one is inlined the other is not), was this left there by mistake?

Reply:
I see, maybe you could make entryCount work like sessionCount?

Reply:
Yes I'll tidy this up afterwards. there is alot more to do here for batch so I'll be refactoring this in due cause.